File: src/main/java/org/hisp/dhis/expression/eval/NodeVisitor.java
package org.hisp.dhis.expression.eval;

import org.hisp.dhis.expression.ast.BinaryOperator;
import org.hisp.dhis.expression.ast.DataItemType;
import org.hisp.dhis.expression.ast.NamedFunction;
import org.hisp.dhis.expression.ast.DataItemModifier;
import org.hisp.dhis.expression.ast.NamedValue;
import org.hisp.dhis.expression.ast.Node;
import org.hisp.dhis.expression.ast.NodeType;
import org.hisp.dhis.expression.ast.UnaryOperator;

import java.time.LocalDateTime;
import java.util.function.Consumer;

/**
 * Extended visitor for {@link Node}s by their {@link NodeType}.
 *
 * @author Jan Bernitt
 */
public interface NodeVisitor extends Consumer<Node<?>>
{
    @Override
    @SuppressWarnings("unchecked")
    default void accept(Node<?> node) {
        switch(node.getType()) {
            // complex nodes
            case UNARY_OPERATOR: visitUnaryOperator((Node<UnaryOperator>) node); break;
            case BINARY_OPERATOR: visitBinaryOperator((Node<BinaryOperator>) node); break;
            case ARGUMENT: visitArgument((Node<Integer>) node); break;
            case PAR: visitParentheses((Node<Void>) node); break;
            case FUNCTION: visitFunction((Node<NamedFunction>) node); break;
            case MODIFIER: visitModifier((Node<DataItemModifier>) node); break;
            case DATA_ITEM: visitDataValue((Node<DataItemType>) node); break;

            // simple nodes
            case BOOLEAN: visitBoolean((Node<Boolean>) node); break;
            case WILDCARD: visitWildcard((Node<Void>) node); break;
            case UID: visitUid((Node<String>) node); break;
            case DATE: visitDate((Node<LocalDateTime>) node); break;
            case NULL: visitNull((Node<Void>) node); break;
            case NUMBER: visitNumber((Node<Double>) node); break;
            case STRING: visitString((Node<String>) node); break;
            case INTEGER: visitInteger((Node<Integer>) node); break;
            case IDENTIFIER: visitIdentifier(node); break;
            case NAMED_VALUE: visitNamedValue((Node<NamedValue>) node); break;

            default: throw new UnsupportedOperationException("Not type not supported yet: "+node.getType());
        }
    }

    default void visitParentheses(Node<Void> group) {
    }

    default void visitArgument(Node<Integer> argument) {

    }

    default void visitBinaryOperator(Node<BinaryOperator> operator) {

    }

    default void visitUnaryOperator(Node<UnaryOperator> operator) {

    }

    default void visitFunction(Node<NamedFunction> function) {

    }

    default void visitModifier(Node<DataItemModifier> modifier) {

    }

    default void visitDataValue(Node<DataItemType> data) {

    }

    /*
    Simple nodes:
     */

    default void visitNamedValue(Node<NamedValue> value)
    {

    }

    default void visitNumber(Node<Double> value) {

    }

    default void visitInteger(Node<Integer> value) {

    }

    default void visitBoolean(Node<Boolean> value)
    {

    }

    default void visitNull(Node<Void> value)
    {

    }

    default void visitWildcard(Node<Void> value)
    {

    }

    default void visitString(Node<String> value)
    {

    }

    default void visitIdentifier(Node<?> value)
    {
        // identifier nodes use both enums or String values so at this point we can't say what value we got
    }

    default void visitUid(Node<String> value) {

    }

    default void visitDate(Node<LocalDateTime> value)
    {

    }
}

File: src/commonMain/kotlin/org/hisp/dhis/lib/expression/eval/NodeVisitor.kt
package org.hisp.dhis.lib.expression.eval

import kotlinx.datetime.LocalDate
import org.hisp.dhis.lib.expression.ast.*
import org.hisp.dhis.lib.expression.spi.DataItemType

/**
 * Extended visitor for [Node]s by their [NodeType].
 *
 * @author Jan Bernitt
 */
@Suppress("kotlin:S1186")
interface NodeVisitor {

    @Suppress("kotlin:S6530", "UNCHECKED_CAST")
    fun visitNode(node: Node<*>) {
        when (node.getType()) {
            NodeType.UNARY_OPERATOR -> visitUnaryOperator(node as Node<UnaryOperator>)
            NodeType.BINARY_OPERATOR -> visitBinaryOperator(node as Node<BinaryOperator>)
            NodeType.ARGUMENT -> visitArgument(node as Node<Int>)
            NodeType.PAR -> visitParentheses(node as Node<Unit>)
            NodeType.FUNCTION -> visitFunction(node as Node<NamedFunction>)
            NodeType.MODIFIER -> visitModifier(node as Node<DataItemModifier>)
            NodeType.DATA_ITEM -> visitDataItem(node as Node<DataItemType>)
            NodeType.VARIABLE -> visitVariable(node as Node<VariableType>)
            NodeType.BOOLEAN -> visitBoolean(node as Node<Boolean>)
            NodeType.UID -> visitUid(node as Node<String>)
            NodeType.DATE -> visitDate(node as Node<LocalDate>)
            NodeType.NULL -> visitNull(node as Node<Unit>)
            NodeType.NUMBER -> visitNumber(node as Node<Double>)
            NodeType.STRING -> visitString(node as Node<String>)
            NodeType.INTEGER -> visitInteger(node as Node<Int>)
            NodeType.IDENTIFIER -> visitIdentifier(node)
            NodeType.NAMED_VALUE -> visitNamedValue(node as Node<NamedValue>)
        }
    }

    fun visitParentheses(group: Node<Unit>) {}
    fun visitArgument(argument: Node<Int>) {}
    fun visitBinaryOperator(operator: Node<BinaryOperator>) {}
    fun visitUnaryOperator(operator: Node<UnaryOperator>) {}
    fun visitFunction(fn: Node<NamedFunction>) {}
    fun visitModifier(modifier: Node<DataItemModifier>) {}
    fun visitDataItem(item: Node<DataItemType>) {}
    fun visitVariable(variable: Node<VariableType>) {}

    /*
    Simple nodes:
     */
    fun visitNamedValue(value: Node<NamedValue>) {}
    fun visitNumber(value: Node<Double>) {}
    fun visitInteger(value: Node<Int>) {}
    fun visitBoolean(value: Node<Boolean>) {}
    fun visitNull(value: Node<Unit>) {}
    fun visitString(value: Node<String>) {}
    fun visitIdentifier(value: Node<*>) {
        // identifier nodes use both enums or String values so at this point we can't say what value we got
    }

    fun visitUid(value: Node<String>) {}
    fun visitDate(value: Node<LocalDate>) {}
}
